lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,377 @@
+import os
+import time
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        if bits in permdict:
+            fperm += permdict[bits]
+        else:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+def header(fil, status, timestamp, hierarchy, path):
+    if status == "start":
+        fil.write(f"\n\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+    else:
+        fil.write(f"\n{status}-{timestamp}: {hierarchy} hierarchy: {path}\n")
+
+
+def dump_check(fil, hierarchy_check, controller, key):
+    if controller in hierarchy_check[key]._dict:
+        fil.write(f"\t{hierarchy_check[key].get_msg()}\n")
+        output_format(fil, hierarchy_check[key]._dict[controller])
+    return 1
+
+
+def output_format(fil, controller_list):
+    for val in sorted(controller_list):
+        fil.write(f"\t\t{val}\n")
+    return 1
+
+
+class Hierarchy(object):
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+
+class DictOfList(object):
+    """Class used to create and add elements to Dictionary
+    """
+
+    def __init__(self, msg):
+        self._dict = OrderedDict()
```

Comment:
Typically in Python, using a leading underscore (`_`) for a field name communicates that the field is a "private", where uses of an instance of that class should not reference that field directly.

You might want to look up examples of Python3 patterns that create dictionary like object behavior, like https://stackoverflow.com/questions/43627405/understanding-getitem-method, which show you how to use the `__getitem__` and `__setitem__` methods of a class to make an object operate like a dictionary would.